Split Drivers and Groups management

Abyss777 opened the pull request:
The main problem is that devices and groups relationship is complex, we use `GroupTree` to trace it.
The next problem is that we have to have list of devices and groups before creating the tree.
I decided to add new layer in object managers structure and it looks next:

- `BaseObjectManager` class that only stores objects cache and do add/update/remove
- `ManagableObjects` (probably need to be renamed) interface that implements two functions `getUserItems` and `getManagedItems`. `getManagedItems` are the same in 3 classes, but it is impossible to use `default` in Java 7. This interface used only to strict the code.
- `SimpleObjectManager` class that implements `ManagableObjects` and store User -> Object permissions
- `ExtendedObjectManager` is not changed, it stores Device -> Object and Group -> Object permissions

I leaved this complex devices and groups tree in `PermissionsManager` and just apply to it from Device and Group managers.

Now objects API resources look very similar and this will allow combine them on some future stage.

Some comments in code.

Abyss777 commented on src/org/traccar/database/PermissionsManager.java:
```diff
@@ -94,11 +94,12 @@ public void refreshServer() {
     }
 
-    public final void refreshPermissions() {
+    public final void refreshDeviceAndGroupPermissions() {
```
Just renamed to reflect what it does.

Abyss777 commented on src/org/traccar/database/PermissionsManager.java:
```diff
@@ -176,9 +177,9 @@ public void checkDeviceLimit(long userId) throws SecurityException, SQLException
+                deviceCount = Context.getDeviceManager().getManagedItems(userId).size();
             } else {
-                deviceCount = getDevicePermissions(userId).size();
+                deviceCount = Context.getDeviceManager().getUserItems(userId).size();
```
Basically nothing changed here, just a couple function calls, but it looks more consistent.

tananaev commented on src/org/traccar/reports/Events.java:
```diff
@@ -92,7 +92,7 @@ public static void getExcel(OutputStream outputStream,
             sheetNames.add(WorkbookUtil.createSafeSheetName(deviceEvents.getDeviceName()));
             if (device.getGroupId() != 0) {
-                Group group = Context.getDeviceManager().getGroupById(device.getGroupId());
+                Group group = (Group) Context.getGroupsManager().getById(device.getGroupId());
```
Can we have a generic function that will automatically return correct object type?

Abyss777 replied:
Similar to `<T> Collection<T> getItems(Class<T> clazz, Set<Long> itemIds)` but for the single object?

tananaev replied:
Yes. Will it work? I think Java should be able to infer the type automatically.

Abyss777 replied:
It will. But it might look longer
````java
Group group = (Group) Context.getGroupsManager().getById(device.getGroupId());
````
vs
````java
Group group = Context.getGroupsManager().getById(Group.class, device.getGroupId());
````

tananaev replied:
I was thinking just
```java
Group group = Context.getGroupsManager().getById(device.getGroupId());
```

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -169,76 +90,129 @@ public Device getDeviceByPhone(String phone) {
-        devicesById.put(device.getId(), device);
-        devicesByUniqueId.put(device.getUniqueId(), device);
+    @Override
+    protected void addNewItem(BaseModel item) {
```
Can we use `Device` directly in the arguments here? Same probably applies to other classes as well.

Abyss777 replied:
I'm afraid we can't. Overridden method have to have same set of parameters

tananaev replied:
Right. We can only change return parameter. Then ignore my comment.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -169,76 +90,129 @@ public Device getDeviceByPhone(String phone) {
+    @Override
+    protected void addNewItem(BaseModel item) {
+        Device device = (Device) item;
+        super.addNewItem(item);
```
Please move super up above variable.

Abyss777 commented:
Overrode `getById` in all managers, except `DeviceManager` because we already have `getDeviceById` for `IdentityManager`

tananaev commented:
Should we just rename methods in `IdentityManager` to match?

tananaev commented on src/org/traccar/database/AttributesManager.java:
```diff
@@ -26,9 +26,14 @@ public AttributesManager(DataManager dataManager) {
     }
 
     @Override
+    public Attribute getById(long calendarId) {
```
I'm thinking now... can't we just change  `ExtendedObjectManager` to be generic?

Abyss777 replied:
It is mistake here, variable should be named `attributeId` of course.

I also thought about it, but I'm not familiar with it, It will take time.

tananaev replied:
It's really easy:

https://docs.oracle.com/javase/tutorial/java/generics/types.html

tananaev commented on src/org/traccar/database/BaseObjectManager.java:
```diff
@@ -0,0 +1,124 @@
+import org.traccar.helper.Log;
+import org.traccar.model.BaseModel;
+
+public class BaseObjectManager {
```
Probably same here. We can make it generic to avoid all those downcasts. What do you think?

Abyss777 commented:
We have
````
Device getDeviceById(long id);

Device getDeviceByUniqueId(String uniqueId) throws Exception;
````
in `IdentityManager` probably we should also rename to `getByUniqueId`

tananaev commented:
Yes, makes sense to rename both.

Abyss777 commented:
Generics are awesome, they perfectly fit to us.  Looks more ideal now. No type casts at all. Hope I've removed them from everywhere.

Also renamed `getDeviceById` to just `getById` a lot of files touched.

tananaev commented on src/org/traccar/database/GroupsManager.java:
```diff
@@ -0,0 +1,104 @@
+                DeviceManager.DEFAULT_REFRESH_DELAY) * 1000;
+    }
+
+    private void checkGroupCycles(BaseModel group) {
```
I think this needs to be updated as well.

Abyss777 commented:
Fixed

tananaev commented:
Merged, thanks.
